src/rust/engine/graph/README.md

```diff
@@ -0,0 +1,49 @@
+# The Runtime Graph
+
+## Overview
+
+Pants' runtime [Graph](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/lib.rs#L547-L572) contains [Node](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/node.rs#L18-L36)s that represent memoized user logic and have a known output type. Edges between `Node`s represent data dependencies, which allow `Node`s to be invalidated when their inputs have changed.
+
+## Nodes
+
+The `Graph` trait is generic for testing purposes, but in Pants' production usage, the `Node` trait [is implemented for](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/src/nodes.rs#L1132-L1139) `@rule`s and intrinsics (essentially, `@rule`s implemented in Rust). The `Node` is a single generic parameter, and so from the `Graph`'s perspective, all `Node` instances are identical: the production usage implements `Node` [using an enum](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/src/nodes.rs#L1030-L1044) in order to isolate data and logic used for different purposes (running `@rule`s, running intrinsics, etc).
+
+### Inputs and Identity
+
+The `Node` trait requires both `Eq` and `Hash`, so a `Node` is its own memoization key and "identity". When an instance of a `Node` runs (via its [`run` method](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/node.rs#L30)), it has two sources of information that it can consume: its own fields/identity, and a [`NodeContext` object](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/node.rs#L66-L117) which allows it to request the values of other `Node`s and access any other information which should not contribute to the `Node`'s identity.
+
+For example: the [`Node` implementation for a `@rule`](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/src/nodes.rs#L936-L996) (called "Task" for historical reasons) holds information that uniquely identifies that instance of the `@rule`: the `Task` struct differentiates it from other `@rule` definitions, and the `Params` (the inputs that will be used to compute its positional arguments and `Get`s) differentiate it from other instances of the same `@rule` definition.
```

Comment:
In a followup, can we purge `Task`? It adds another layer of cognitive load to our Rust code.